src/glewlwyd.c

```diff
@@ -181,15 +183,27 @@ int main (int argc, char ** argv) {
       pthread_cond_init(&global_handler_close_cond, NULL)) {
     y_log_message(Y_LOG_LEVEL_ERROR, "init - Error initializing global_handler_close_lock or global_handler_close_cond");
   }
-  // Catch end signals to make a clean exit
-  if (signal (SIGQUIT, exit_handler) == SIG_ERR ||
-      signal (SIGINT, exit_handler) == SIG_ERR ||
-      signal (SIGTERM, exit_handler) == SIG_ERR ||
-      signal (SIGHUP, exit_handler) == SIG_ERR ||
-      signal (SIGBUS, exit_handler) == SIG_ERR ||
-      signal (SIGSEGV, exit_handler) == SIG_ERR ||
-      signal (SIGILL, exit_handler) == SIG_ERR) {
-    fprintf(stderr, "init - Error initializing end signal\n");
+
+  // Process end signals on dedicated thread
+  if (sigemptyset(&close_signals) == -1 ||
+      sigaddset(&close_signals, SIGQUIT) == -1 ||
+      sigaddset(&close_signals, SIGINT) == -1 ||
+      sigaddset(&close_signals, SIGTERM) == -1 ||
+      sigaddset(&close_signals, SIGHUP) == -1 ||
+      sigaddset(&close_signals, SIGBUS) == -1 ||
+      sigaddset(&close_signals, SIGSEGV) == -1 ||
+      sigaddset(&close_signals, SIGILL) == -1) {
+    fprintf(stderr, "init - Error creating signal mask\n");
+    exit_server(&config, GLEWLWYD_ERROR);
+  }
+  if (pthread_sigmask(SIG_BLOCK, &close_signals, NULL)) {
+    fprintf(stderr, "init - Error setting signal mask\n");
+    exit_server(&config, GLEWLWYD_ERROR);
+  }
+
+  if (pthread_create(&signal_thread_id, NULL, &signal_thread, &close_signals)) {
```

Comment:
Also, I think we should detach the thread to release its resource when it finishes.

```C
if (pthread_create(&signal_thread_id, NULL, &signal_thread, &close_signals) || pthread_detach(&signal_thread_id)) {
```

Reply:
no, apparently you don't need to detach this thread, my bad